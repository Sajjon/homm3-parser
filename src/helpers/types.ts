export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export type FlaggedProp<FlagName extends string, ValueName extends string, T> =
  | Record<FlagName, true> & Record<ValueName, T>
  | Record<FlagName, false>
