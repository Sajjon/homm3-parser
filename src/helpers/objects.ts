import { ObjectType, objectTypeToMeta } from '../parsers/h3m/enums/object'
import { MetaType } from '../parsers/h3m/contracts/enums/MetaType'
import { createAdapter, TagProducer, TagOrWrapper } from 'binary-markup'

export function getMetaType(type: ObjectType): MetaType {
  if (type in objectTypeToMeta) {
    return objectTypeToMeta[type]
  }
  return MetaType.None
}

export const bitMasksArray = <T>(
  tag: TagOrWrapper<number[]>,
  bitGroups: [number, T][][],
): TagProducer<T[]> =>
  createAdapter<number[], T[]>(
    tag,
    (data: number[]) => {
      return data.reduce((acc: T[], mask, index) => {
        const bits = bitGroups[index]
        for (const [bit, skill] of bits) {
          if (mask & bit) {
            acc.push(skill)
          }
        }
        return acc
      }, [])
    },
    (skills: T[]) => {
      return bitGroups.map(bits => {
        let mask = 0
        for (const [bit, skill] of bits) {
          if (skills.includes(skill)) {
            mask |= bit
          }
        }
        return mask
      })
    },
  )
