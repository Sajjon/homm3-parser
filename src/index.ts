import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import * as pako from 'pako'
import { h3mFile } from './parsers/h3m/h3m.bml'
import { H3MFile } from './parsers/h3m/contracts'

export * from './parsers/assets'
export * from './parsers/h3m/contracts'
export * from './parsers/h3m/constants'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = pako.inflate(data)
  return parseH3M(content.buffer)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
