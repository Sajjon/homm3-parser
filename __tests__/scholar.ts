import { testMap } from './helpers/utils'
import { H3MFile } from '../src'

describe('scholar', () => {
  testMap('scholar', 'scholar.h3m', (map: H3MFile) => map.objectDetails)
})
