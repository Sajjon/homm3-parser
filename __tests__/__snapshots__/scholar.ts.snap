// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scholar should parse scholar 1`] = `
Object {
  "count": 4,
  "entries": Array [
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVXschl0.def",
        "landscapeGroup": 255,
        "metaType": "Scholar",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "Scholar",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "rewardType": "Random",
          "rewardValue": 0,
        },
        "type": "Scholar",
      },
      "oaIndex": 2,
      "x": 1,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVXschl0.def",
        "landscapeGroup": 255,
        "metaType": "Scholar",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "Scholar",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "rewardType": "PrimarySkill",
          "rewardValue": "SpellPower",
        },
        "type": "Scholar",
      },
      "oaIndex": 2,
      "x": 3,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVXschl0.def",
        "landscapeGroup": 255,
        "metaType": "Scholar",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "Scholar",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "rewardType": "SecondarySkill",
          "rewardValue": "Wisdom",
        },
        "type": "Scholar",
      },
      "oaIndex": 2,
      "x": 5,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVXschl0.def",
        "landscapeGroup": 255,
        "metaType": "Scholar",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "Scholar",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "rewardType": "Spell",
          "rewardValue": "Prayer",
        },
        "type": "Scholar",
      },
      "oaIndex": 2,
      "x": 7,
      "y": 1,
      "z": 0,
    },
  ],
}
`;
